src/shims/sync.rs

```diff
@@ -435,17 +434,23 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
         let id = mutex_get_or_create_id(this, mutex_op)?;
+        let active_thread = this.get_active_thread()?;
 
-        if let Some((old_owner_thread, _old_locked_count)) = this.mutex_unlock(id)? {
-            // The mutex was locked by someone (`owner_thread`).
-            if old_owner_thread != this.get_active_thread()? {
-                throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
-            }
+        if let Some(_old_locked_count) = this.mutex_unlock(id, active_thread)? {
+            // The mutex was locked by the current thread.
             Ok(0)
         } else {
-            // The mutex was not locked by anyone.
-            if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
-                throw_ub_format!("unlocked a PTHREAD_MUTEX_NORMAL mutex that was not locked");
+            // The mutex was locked by another thread or not locked at all. See
+            // the “Unlock When Not Owner” column in
+            // https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_unlock.html.
+            if kind == this.eval_libc("PTHREAD_MUTEX_DEFAULT")? {
```

Comment:
None of our other `if else` chain also considers `PTHREAD_MUTEX_DEFAULT`, or do they?

Reply:
Now, after fixing #1419, they do.